File: src/1268-Search-Suggestions-System/1268.cpp
class Solution 
{
public:
    vector<vector<string>> suggestedProducts(vector<string>& products, string searchWord) 
    {
        sort(products.begin(), products.end());
        vector<vector<string>> res;
        int l = 0, r = products.size() - 1;
        
        for (int i = 0; i < searchWord.size(); i++)
        {
            while (l <= r && (products[l].size() <= i || products[l][i] != searchWord[i])) l++;
            while (l <= r && (products[r].size() <= i || products[r][i] != searchWord[i])) r--;
            if (l <= r) res.push_back(vector<string>(products.begin() + l, products.begin() + min(l+3, r+1)));
            else res.push_back({});
        }
        return res;
    }
};

File: src/1268-Search-Suggestions-System/1268.py
class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        products.sort()
        l, r, res = 0, len(products) - 1, []
        for i, c in enumerate(searchWord):
            l = bisect.bisect_left(products, searchWord[:i+1], l, r + 1)
            while l <= r and (len(products[r]) <= i or products[r][i] != c):
                r -= 1
            res.append(products[l:min(l + 3, r + 1)] if l <= r else [])
        return res